File: src/main/java/net/katsstuff/danmakucore/impl/shape/ShapeEffectDelay.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.impl.shape;

import java.util.Set;

import net.katsstuff.danmakucore.data.Quat;
import net.katsstuff.danmakucore.data.Vector3;
import net.katsstuff.danmakucore.entity.danmaku.EntityDanmaku;
import net.katsstuff.danmakucore.shape.IShape;
import net.katsstuff.danmakucore.shape.ShapeResult;

public class ShapeEffectDelay implements IShape {

	private final IShape shape;
	private final int delay;

	public ShapeEffectDelay(IShape shape, int delay) {
		this.shape = shape;
		this.delay = delay;
	}

	@Override
	public ShapeResult draw(Vector3 pos, Quat orientation, int tick) {
		ShapeResult res = shape.draw(pos, orientation, tick);
		if(tick < delay) return ShapeResult.notDone(res.getSpawnedDanmaku());
		else return res;
	}

	@Override
	public void doEffects(Vector3 pos, Quat orientation, int tick, Set<EntityDanmaku> spawnedThisTick, Set<EntityDanmaku> allSpawned) {
		if(tick >= delay) {
			shape.doEffects(pos, orientation, tick - delay, spawnedThisTick, allSpawned);
		}
	}
}

File: src/main/scala/net/katsstuff/danmakucore/impl/shape/ShapeEffectDelay.scala
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.impl.shape

import net.katsstuff.mirror.data.{Quat, Vector3}
import net.katsstuff.danmakucore.shape.{Shape, ShapeResult}

class ShapeEffectDelay(val shape: Shape, val delay: Int) extends Shape {
  override def draw(pos: Vector3, orientation: Quat, tick: Int): ShapeResult = {
    val res = shape.draw(pos, orientation, tick)
    if (tick < delay) ShapeResult.notDone(res.spawnedDanmaku)
    else res
  }

  override def doEffects(
      pos: Vector3,
      orientation: Quat,
      tick: Int,
      spawnedThisTick: ShapeResult,
      allSpawned: Set[ShapeResult]
  ): Unit = if (tick >= delay) shape.doEffects(pos, orientation, tick - delay, spawnedThisTick, allSpawned)
}
